pkg/didcomm/crypto/legacy/authcrypt/authcrypt.go

```diff
@@ -0,0 +1,200 @@
+}
+
+func isKeyPairValid(kp keyPairEd25519) bool {
+	if kp.priv == nil || kp.pub == nil {
+		return false
+	}
+
+	return true
+}
+
+// envelope is the full payload envelope for the JSON message
+type envelope struct {
+	Protected  string `json:"protected,omitempty"`
+	IV         string `json:"iv,omitempty"`
+	CipherText string `json:"ciphertext,omitempty"`
+	Tag        string `json:"tag,omitempty"`
+}
+
+// protected is the protected header of the JSON envelope
+type protected struct {
+	Enc        string      `json:"enc,omitempty"`
+	Typ        string      `json:"typ,omitempty"`
+	Alg        string      `json:"alg,omitempty"`
+	Recipients []recipient `json:"recipients,omitempty"`
+}
+
+// recipient holds the data for a recipient in the envelope header
+type recipient struct {
+	EncryptedKey string          `json:"encrypted_key,omitempty"`
+	Header       recipientHeader `json:"header,omitempty"`
+}
+
+// recipientHeader holds the header data for a recipient
+type recipientHeader struct {
+	KID    string `json:"kid,omitempty"`
+	Sender string `json:"sender,omitempty"`
+	IV     string `json:"iv,omitempty"`
+}
+
+// publicEd25519toCurve25519 takes an Ed25519 public key and provides the corresponding Curve25519 public key
+//  This function wraps PublicKeyToCurve25519 from Adam Langley's ed25519 repo: github.com/agl/ed25519
+func publicEd25519toCurve25519(pub *publicEd25519) (*publicCurve25519, error) {
+	pkOut := new([CurveKeySize]byte)
+	success := extra25519.PublicKeyToCurve25519(pkOut, (*[CurveKeySize]byte)(pub))
+	if !success {
+		return nil, errors.New("failed to convert public key")
+	}
+	return (*publicCurve25519)(pkOut), nil
+}
+
+// TODO: consider wrapping agl's secret key conversion instead of reimplementing
+
+// secretEd25519toCurve25519 converts a secret key from Ed25519 to curve25519 format
+// Made with reference to https://github.com/agl/ed25519/blob/master/extra25519/extra25519.go and
+// https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_sign/ed25519/ref10/keypair.c#L70
+func secretEd25519toCurve25519(priv *privateEd25519) (*privateCurve25519, error) {
+	hasher := sha512.New()
+	_, err := hasher.Write(priv[:32])
+	if err != nil {
+		return nil, err
+	}
+
+	hash := hasher.Sum(nil)
+
+	hash[0] &= 248  // clr lower 3 bits
+	hash[31] &= 127 // clr upper 1 bit
+	hash[31] |= 64  // set 6th bit
+
+	out := new([CurveKeySize]byte)
+	copy(out[:], hash)
+	return (*privateCurve25519)(out), nil
+}
+
+func makeNonce(pub1, pub2 []byte) ([]byte, error) {
+	var nonce [24]byte
+	// generate an equivalent nonce to libsodium's (see link above)
+	nonceWriter, err := blake2b.New(24, nil)
+	if err != nil {
+		return nil, err
+	}
+	_, err = nonceWriter.Write(pub1)
+	if err != nil {
+		return nil, err
+	}
+	_, err = nonceWriter.Write(pub2)
+	if err != nil {
+		return nil, err
+	}
+
+	nonceOut := nonceWriter.Sum(nil)
+	copy(nonce[:], nonceOut)
+
+	return nonce[:], nil
+}
+
+// TODO dupe of jwe/authcrypt encryptOID, refactor out
```

Comment:
most probably this function will stay here as jwe/authcrypt encryptOID will eventually be replaced with a JWK field (and this function removed from there).

Reply:
Will remove TODO